src/zope/interface/registry.py

```diff
@@ -38,6 +41,129 @@ def notify(*arg, **kw): pass
 from zope.interface._compat import STRING_TYPES
 
 
+class _UnhashableComponentCounter(object):
+    # defaultdict(int)-like object for unhashable components
+
+    def __init__(self, otherdict):
+        # [(component, count)]
+        self._data = [item for item in otherdict.items()]
+
+    def __getitem__(self, key):
+        for component, count in self._data:
+            if component == key:
+                return count
+        return 0
+
+    def __setitem__(self, component, count):
+        for i, data in enumerate(self._data):
+            if data[0] == component:
+                self._data[i] = component, count
+                return
+        self._data.append((component, count))
+
+    def __delitem__(self, component):
+        for i, data in enumerate(self._data):
+            if data[0] == component:
+                del self._data[i]
+                return
+        raise KeyError(component) # pragma: no cover
+
+
+class _UtilityRegistrations(object):
+
+    _regs_for_components = WeakKeyDictionary()
```

Comment:
Why is this connected to utility registries via a WeakKeyDictionary rather than just adding a ``_cache`` attribute to Components?  This indirection seems pretty weird because it distributes the state.